pages/categorization.vue

```diff
@@ -0,0 +1,521 @@
+<template>
+
+	<b-container fluid>
+
+		<tool-navbar :navItemsState="navItemsState" :pageName="fullName"></tool-navbar>
+
+		<b-row>
+
+			<b-col cols="4">
+				<coloring-listgroup
+					:columnData="recommendedColumns"
+					:defaultPalette="$store.state.pageData.categorization.default"
+					tag="recommended-column"
+					title="Recommended Columns"
+					instructions="Click column type and then corresponding column from tsv file"
+					v-on:paint-action="$store.dispatch('saveCurrentPaintInfo', $event)"
+					>
+				</coloring-listgroup>
+			</b-col>
+
+			<b-col cols="8">
+				<filedata-table
+					:fields="columnMatchTableFields"
+					:paintClass="getPaintClass"
+					:currentPalette="$store.state.pageData.categorization.current"
+					:defaultPalette="$store.state.pageData.categorization.default"
+					:tableData="tableDataFromTsvAndOrJson"
+					v-on:column-name-selected="tableClick($event)">
+				</filedata-table>
+			</b-col>
+
+		</b-row>
+
+		<b-row>
+			
+			<b-col cols="9"></b-col>
+			
+			<b-col cols="3">
+				<!-- Only enabled when at least one column has been categorized -->
+				<b-button
+					class="float-right"
+					:disabled="!readyForNextStepFlag"
+					:to="'/' + pageNames.annotation.location"
+					:variant="nextPageButtonColor">
+					Next step: Annotate columns
+				</b-button>
+			</b-col>
+			
+		</b-row>		
+
+	</b-container>
+
+</template>
+
+<script>
+
+	// String literals
+
+	export default {
+
+		name: "CategorizationPage",
+
+		created() {
+
+			// Determine page state from data contents and change to that new state
+			this.changeToNewState();
+		},		
+
+		data() {
+
+			return {
+
+				// Columns for file data table	
+				columnMatchTableFields: [
+
+					{ key: "column" },
+					{ key: "description" }
+				],
+
+				// Current state of the page
+				currentState: 0,
+
+				// Default table background color
+				defaultBackgroundColor: this.$store.getters.pageData.categorization.default.bColor,
+
+				// Full text name of this page
+				fullName: this.$store.getters.pageNames.categorization.fullName, 
+
+				// Status of the navbar item links for other pages
+				navItemsState: [
+					
+					{
+						enabled: true,
+						pageInfo: this.$store.state.pageNames.home
+					},
+					{
+						enabled: false,
+						pageInfo: this.$store.state.pageNames.annotation
+					},
+					{
+						enabled: false,
+						pageInfo: this.$store.state.pageNames.download
+					}
+				],
+
+				// Bootstrap variant color of the button leading to the categorization page
+				nextPageButtonColor: "secondary",
+
+				// Possible states of this page
+				possibleStates: {
+
+					STATE_NOCATEGORIES_PAINTED: 0,
+					STATE_ATLEASTONE_CATEGORY_PAINTED: 1 << 1
```

Comment:
Check out my explanation in the page_states.md file I wrote yesterday. 1 << 0 == 1,  1 << 1 == 2, 1 << 2 == 4.
The idea here is to have a number where each bit represents a unique flag. So it totally could be written as 1, 2, 4, etc.

(Small issue here is I began with 1 << 1, where I should have begun with 1 << 0)

Reply:
At the risk of turning this into a JS lesson just for myself, I'd like to understand this better first. The page can have states. If I understand this right, the two possible states are 
- `STATE_NOCATEGORIES_PAINTED` and 
- `STATE_ATLEASTONE_CATEGORY_PAINTED`. 

is this not just a boolean thing? I.e. wouldn't it be sufficient if the values for these two keys are either `true` or `false`? 
I think these are the notes you mentioned: 
> This object variable stores the list of bit flags representing each state. These are meant to be unique binary numbers (e.g. 1, 2, 4, 8, etc.) with 0 representing 'no state'. Values can either be set plainly via decimal numbers or, if it's helpful to be more explicit, via the bit-shift operator (e.g. '1 << 1', '1 << 2', '1 << 3', '1 << 4

So I guess my questions would be here:
- Why do we need unique flags for the `STATE_ATLEASTONE_CATEGORY_PAINTED`. If I understand this correctly, `possibleStates` is not shared between pages.
- If we do need unique flags, why do they have to be numeric? Would it be possible to use strings instead? and if not, we should add detailed comments to explain what each numeric flag means.

I think this is a two-part question. 
1. Can we make the state encoding easier via boolean / intuitively understandable keywords?
2. Is this bitwise shift syntax a widely used thing in JS world so that we can expect even a relative novice to be familiar with it.

Reply:
Hey @jarmoza, I have been giving this a bit more of a think. I understand the point behind this now and I do agree it is an elegant solution. But I think based on what we need so far, I vote to go with boolean flags and come back to this if and when things get unwieldy. For context, we are using this bit field flag style in 2 places, for a total of 6 states:

- `categorization.vue`
  - STATE_NOCATEGORIES_PAINTED
  - STATE_ATLEASTONE_CATEGORY_PAINTED
- `index.vue`
  - `STATE_NOFILES_LOADED`
  - `STATE_TSVFILE_LOADED`
  - `STATE_JSONFILE_LOADED`
  - `STATE_BOTHFILES_LOADED`

This can be done by:
- `categorization.vue`
  - `STATE_CATEGORY_PAINTED` = oneof(True, False)
- `index.vue`
  - `this.$store.state.pageData.home.tsvFile != null`
  - `this.$store.state.pageData.home.jsonFile`
  - (in fact this is how we are using it in `categorization.vue` already

The main benefit here is legibility, and how easy it is for us to come back and understand what's going on in the future. When we find that this is not performant enough at some point, we know how to fix it then. 

Does that sound OK to you?